File: examples/java_project/src/main/java/com/example/StringUtils.java
package com.example;

import java.util.ArrayList;
import java.util.List;

/**
 * Simple string utility class for testing CodeMorph Java to Python translation.
 */
public class StringUtils {

    // Constants
    public static final int MAX_LENGTH = 1000;
    public static final String DEFAULT_SEPARATOR = ",";

    /**
     * Reverse a string.
     *
     * @param input The string to reverse
     * @return The reversed string
     */
    public static String reverse(String input) {
        if (input == null) {
            return null;
        }
        StringBuilder sb = new StringBuilder(input);
        return sb.reverse().toString();
    }

    /**
     * Check if a string is a palindrome.
     *
     * @param input The string to check
     * @return true if the string is a palindrome
     */
    public static boolean isPalindrome(String input) {
        if (input == null || input.isEmpty()) {
            return false;
        }
        String cleaned = input.toLowerCase().replaceAll("[^a-z0-9]", "");
        String reversed = reverse(cleaned);
        return cleaned.equals(reversed);
    }

    /**
     * Count occurrences of a character in a string.
     *
     * @param input The string to search
     * @param target The character to count
     * @return The number of occurrences
     */
    public static int countChar(String input, char target) {
        if (input == null) {
            return 0;
        }
        int count = 0;
        for (char c : input.toCharArray()) {
            if (c == target) {
                count++;
            }
        }
        return count;
    }

    /**
     * Split a string by a separator and return as list.
     *
     * @param input The string to split
     * @param separator The separator
     * @return List of substrings
     */
    public static List<String> splitToList(String input, String separator) {
        List<String> result = new ArrayList<>();
        if (input == null || input.isEmpty()) {
            return result;
        }
        String[] parts = input.split(separator);
        for (String part : parts) {
            String trimmed = part.trim();
            if (!trimmed.isEmpty()) {
                result.add(trimmed);
            }
        }
        return result;
    }

    /**
     * Join a list of strings with a separator.
     *
     * @param items The list of strings to join
     * @param separator The separator to use
     * @return The joined string
     */
    public static String join(List<String> items, String separator) {
        if (items == null || items.isEmpty()) {
            return "";
        }
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < items.size(); i++) {
            sb.append(items.get(i));
            if (i < items.size() - 1) {
                sb.append(separator);
            }
        }
        return sb.toString();
    }

    /**
     * Truncate a string to a maximum length.
     *
     * @param input The string to truncate
     * @param maxLength Maximum length
     * @return Truncated string with ellipsis if needed
     * @throws IllegalArgumentException if maxLength is negative
     */
    public static String truncate(String input, int maxLength) {
        if (maxLength < 0) {
            throw new IllegalArgumentException("maxLength cannot be negative");
        }
        if (input == null || input.length() <= maxLength) {
            return input;
        }
        if (maxLength <= 3) {
            return input.substring(0, maxLength);
        }
        return input.substring(0, maxLength - 3) + "...";
    }
}

File: examples/java_project_translated/string_utils.py
"""
Simple string utility module for testing CodeMorph Java to Python translation.

Translated from: com/example/StringUtils.java
"""

import re
from typing import Optional

# Constants
MAX_LENGTH: int = 1000
DEFAULT_SEPARATOR: str = ","


def reverse(input_str: Optional[str]) -> Optional[str]:
    """
    Reverse a string.

    Args:
        input_str: The string to reverse

    Returns:
        The reversed string
    """
    if input_str is None:
        return None
    return input_str[::-1]


def is_palindrome(input_str: Optional[str]) -> bool:
    """
    Check if a string is a palindrome.

    Args:
        input_str: The string to check

    Returns:
        True if the string is a palindrome
    """
    if input_str is None or input_str == "":
        return False
    cleaned = re.sub(r'[^a-z0-9]', '', input_str.lower())
    reversed_str = reverse(cleaned)
    return cleaned == reversed_str


def count_char(input_str: Optional[str], target: str) -> int:
    """
    Count occurrences of a character in a string.

    Args:
        input_str: The string to search
        target: The character to count

    Returns:
        The number of occurrences
    """
    if input_str is None:
        return 0
    count = 0
    for c in input_str:
        if c == target:
            count += 1
    return count


def split_to_list(input_str: Optional[str], separator: str) -> list[str]:
    """
    Split a string by a separator and return as list.

    Args:
        input_str: The string to split
        separator: The separator

    Returns:
        List of substrings
    """
    result: list[str] = []
    if input_str is None or input_str == "":
        return result
    parts = input_str.split(separator)
    for part in parts:
        trimmed = part.strip()
        if trimmed:
            result.append(trimmed)
    return result


def join(items: Optional[list[str]], separator: str) -> str:
    """
    Join a list of strings with a separator.

    Args:
        items: The list of strings to join
        separator: The separator to use

    Returns:
        The joined string
    """
    if items is None or len(items) == 0:
        return ""
    return separator.join(items)


def truncate(input_str: Optional[str], max_length: int) -> Optional[str]:
    """
    Truncate a string to a maximum length.

    Args:
        input_str: The string to truncate
        max_length: Maximum length

    Returns:
        Truncated string with ellipsis if needed

    Raises:
        ValueError: if max_length is negative
    """
    if max_length < 0:
        raise ValueError("max_length cannot be negative")
    if input_str is None or len(input_str) <= max_length:
        return input_str
    if max_length <= 3:
        return input_str[:max_length]
    return input_str[:max_length - 3] + "..."


if __name__ == "__main__":
    # Simple test
    print(f"reverse('hello') = {reverse('hello')}")
    print(f"is_palindrome('racecar') = {is_palindrome('racecar')}")
    print(f"count_char('hello world', 'l') = {count_char('hello world', 'l')}")
    print(f"split_to_list('a, b, c', ',') = {split_to_list('a, b, c', ',')}")
    print(f"join(['a', 'b', 'c'], '-') = {join(['a', 'b', 'c'], '-')}")
    print(f"truncate('hello world', 8) = {truncate('hello world', 8)}")
